spec-0007/index.md

```diff
@@ -0,0 +1,95 @@
+
+<!--
+Briefly and clearly describe the proposal.
+Explain the general need and the advantages of this specific proposal.
+If relevant, include examples of how the new functionality would be used,
+intended use-cases, and pseudo-code illustrating its use.
+-->
+
+There is disparity in the APIs libraries use to seed random number generation.
+The goal of this SPEC is suggest a single, pragmatic API for the ecosystem, taking into account technical and historical factors.
+
+### Concepts
+
+- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`np.random.default_rng`) uses PCG64.
+- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
+- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.
+
+### Constraints
+
+NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
+Common keyword arguments include `random_state` and `seed`.
+In practice, the seed is unfortunately also often controlled using `np.random.seed`.
+
+## Implementation
+
+<!--
+Discuss how this would be implemented.
+-->
+
+Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.
+
+Two strong motivations for moving over to `Generator`s are:
+
+(1) it avoids naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
+(2) it avoids using global state for seeding.
+
+Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen[^hinsen] principle.
+
+The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:
+
+1. Accept both `rng` and `random_state` keyword arguments.
+2. If `rng=None`, handle `random_state` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.
+   A DeprecationWarning is raised to warn about a future change in behavior.
+3. After <X time>, use only `rng`, seeding with `default_rng(rng)`.
+   Raise an error if `random_state` is provided.
+4. At a time of the lirbrary's choosing, remove any machinery related to `random_state`.
+
+### Impact
+
+The following users will be affected:
+
+1. Those who use `np.random.seed`. The proposal will do away with that global seeding mechanism, meaning that code that relies on it will, after a certain deprecation period, start seeing a different stream of random numbers than before.
+
+   Such code will, in effect, go from being seeded to being unseeded.
+   To avoid that from happening, the code will have to be modified to pass in explicitly an `rng` argument on each function call.
```

Comment:
This change should give a `FutureWarning` if the user called `np.random.seed()` at any time.  If they did not call `np.random.seed()` then the warning is unnecessary because the stream is not reproducible.

Reply:
OK, if NumPy can commit to that, that'd be great! I've added it in.

Reply:
I think I misunderstood what you meant; you meant the decorator should do this check and warn?

Reply:
Yes, sorry, I tried to fix it a bit in the PR (will update now for the other changes).  However, that code does nevertheless reach into NumPy internals, so there is some amount of "commitment" to not break it.